repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsRetryableReadChannel.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2020 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.gcs;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.opensearch.core.internal.io.IOUtils;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.RestorableState;
+import com.google.cloud.RetryHelper;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.StorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Wrapper around reads from GCS that will retry blob downloads that fail part-way through, resuming from where the 
+ * failure occurred. It uses the fact that ReadChannel supports capture and restore functionality and could be recreated
+ * from the last successful checkpoint.
+ */
+public class GcsRetryableReadChannel implements ReadableByteChannel {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsRetryableReadChannel.class);
+    private static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private volatile ReadChannel delegate;
```

Comment:
No need to use `volatile` `GcsRetryableReadChannel` is thread safe. It needs to be `final`. :-)